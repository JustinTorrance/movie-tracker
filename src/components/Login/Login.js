import React, { Component } from 'react'
import { fetchData } from '../../utils/apiCalls'
import { Redirect } from 'react-router-dom'

export default class Login extends Component {
  constructor() {
    super()
    this.state = {
      email: '',
      password: '',
      validUser: false,
      incorrectPassword: false
    }
  }

  handleChange = (e) => {
    const { value, name } = e.target
    this.setState({
      [name]: value
    })
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    const { email, password } = this.state
    const users = await fetchData('http://localhost:3000/api/users')
    let validUser = false

    users.data.find(user => {
      if (user.email === email && user.password === password) validUser = true
    })
    this.setState({validUser, incorrectPassword: !validUser})
  }


  render() {
    if(this.state.validUser) {
      return (
        <Redirect to='/' />
      )
    }
    const { email, password, incorrectPassword } = this.state
    return (
      <form onSubmit={this.handleSubmit}>
        <input 
          className={`form-input ${incorrectPassword && 'incorrect'}`}
          type='text' 
          placeholder='email' 
          value={email} 
          name='email'
          onChange={this.handleChange}>
        </input>
        <input 
          className={`form-input ${incorrectPassword && 'incorrect'}`}
          type='password' 
          placeholder='password' 
          value={password}
          name='password'
          onChange={this.handleChange}>
        </input>
        <button>Submit</button>
      </form>
    )
  }
}