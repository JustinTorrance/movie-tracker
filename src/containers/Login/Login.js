import React, { Component } from 'react'
import { fetchData } from '../../utils/apiCalls'
import { Redirect, Link } from 'react-router-dom'
import { signIn } from '../../actions/index.js'
import { connect } from 'react-redux'
import { PropTypes } from 'prop-types'
import { loginUser } from  '../../utils/apiCalls'

export class Login extends Component {
  constructor() {
    super()
    this.state = {
      email: '',
      password: '',
      validUser: false,
      incorrectLogin: false
    }
  }

  handleChange = (e) => {
    const { value, name } = e.target
    this.setState({
      [name]: value
    })
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    const { email, password } = this.state
    try {
      const user = await loginUser({email, password})
      this.props.loginUser({name: user.data.name, password: user.data.password})
      this.setState({validUser: true})
    } catch (error) {
      this.setState({incorrectLogin: true})
    }

    // const users = await fetchData('http://localhost:3000/api/users')
    // let validUser = false
    // const currentUser = users.data.find(user => {
    //   if (user.email === email && user.password === password) {
    //     validUser = true
    //     return user
    //   } 
    // })
    
    // if (validUser) {
    //   this.props.loginUser(currentUser)
    // }
    // this.setState({validUser, incorrectLogin: !validUser})
  }

  render() {
    if(this.state.validUser) {
      return (
        <Redirect to='/' />
      )
    }
    const { email, password, incorrectLogin } = this.state
    return (
      <div>
        <h1 className='form-title'>
          movie
          <img src='./images/movie_roll.svg' alt="movie roll" className='form-logo'></img>
          tracker
        </h1>
        <h2 className='login-title'>Login</h2>
        <form className='login-form' onSubmit={this.handleSubmit}>
          <input 
            className='form-input'
            type='text' 
            placeholder='email' 
            value={email} 
            name='email'
            onChange={this.handleChange}>
          </input>
          <input 
            className='form-input'
            type='password' 
            placeholder='password' 
            value={password}
            name='password'
            onChange={this.handleChange}>
          </input>
          <button>Login</button>
        </form>
        <div className='new-account-container'>
          <h3 className='incorrect-login'>username or password is incorrect</h3>
          <h3>Don't have an account?</h3>
          <Link className='form-link' to='/signup' >Create New Account</Link>
        </div>
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({ 
  loginUser: (user) => dispatch(signIn(user))
})

Login.propTypes = {
  loginUser: PropTypes.func.isRequired
}

export default connect(null, mapDispatchToProps)(Login)