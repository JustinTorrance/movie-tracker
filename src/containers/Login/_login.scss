::placeholder {
  color: rgb(182, 182, 182);
}

.form-title {
  color: $color-blue;
  font-family: $title-font;
  font-size: 3.8em;
  width: 400px;
  margin: 20px auto;
  text-align: center;
}

.form-logo {
  height: 40px;
  position: relative;
  bottom: -3px;
  width: 40px;
}

.login-title {
  font-family: $regular-font;
  font-weight: 300;
  color: white;
  font-size: 4em;
  width: 400px;
  margin: 20px auto;
  text-align: center;
}

.login-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 370px;
  margin: 10px auto;

  input {
    background-color: rgb(68, 68, 68);
    color: white;
    margin: 10px;
    height: 30px;
    border: 2px solid $color-blue;
    border-radius: 4px;
    font-family: $regular-font;
    font-size: 1.3em;
    font-weight: 300;
    text-align: center;
    width: 100%;
  }

  .incorrect {
    background-color: rgb(187, 95, 95);
    color: white;
  }

  button {
    background-color: rgb(42, 133, 168);
    color: white;
    border: 2px solid rgb(68, 68, 68);
    border-radius: 4px;
    height: 40px;
    margin: 15px auto 0px;
    font-family: $regular-font;
    font-size: 1.3em;
    width: 101%;
    transition: all .4s ease;

    &:hover {
      cursor: pointer;
      color: $color-blue;
      background-color: rgb(48, 48, 48);
      transform: scale(1.01)
    }
  }
}

.new-account-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  font-family: $regular-font;
  font-size: 1.2em;
  font-weight: 300;
  width: 320px;
  margin: 20px auto;

  h3 {
    text-align: center;
    color: white;
    margin: 5px;
  }

  .incorrect-login {
    display: none;
    font-weight: 300;
    font-size: 1em;
    margin: 15px auto;
  }
  
  .incorrect {
    display: block;
    color: rgb(231, 40, 40)
  }
}

.form-link {
  color: $color-blue;
  text-align: center;
  transition: all .4s ease;
  text-decoration: none;
  font-family: $regular-font;
  font-size: 1.2em;
  font-weight: 300;
  width: 320px;
  margin: 20px auto;

  &:visited {
    color: $color-blue
  }

  &:hover {
    text-decoration: underline;
  }
}