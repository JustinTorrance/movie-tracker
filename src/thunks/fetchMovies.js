import { loadMovies } from '../actions/index'
import { loading } from '../actions/index'
import { catchError } from '../actions/index'


export const fetchMovies = (url) => {
  return async (dispatch) => {
    try {
      dispatch(loading(true))
      const response = await fetch(url)
      const movies = await response.json()
      dispatch(loadMovies(movies))
      dispatch(loading(false))      
    } catch(error) {
      dispatch(catchError(true))
      return 'Error: something went wrong'
    }
  }
}