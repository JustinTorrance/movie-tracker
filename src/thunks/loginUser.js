import { signIn } from '../actions/index'

export const loginUser = (user) => {
  return async (dispatch) => {
    try {
      const response = await fetch("http://localhost:3000/api/users/", {
        method: 'POST',
        body: JSON.stringify(user),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      const data = await response.json()
      dispatch(signIn(data.data))
    } catch(error) {
      console.log('error')
      throw (new Error())
    }
  }
}