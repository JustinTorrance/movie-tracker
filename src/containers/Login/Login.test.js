import { Login, mapDispatchToProps } from './Login'
import React from 'react'
import { shallow } from 'enzyme'
import { loginUser } from '../../thunks/loginUser'

jest.mock('../../thunks/loginUser')
loginUser.mockImplementation((data) => {
  return data
})

describe('Login Component', () => {
  let loginUserMock;
  let wrapper;
  
  beforeEach(() => {
    loginUserMock = jest.fn()
    wrapper = shallow(<Login loginUser={loginUserMock} />)
  })
  
  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  
  it('should match the snapshot when user is logged in', () => {
    wrapper.setState({validUser: true})
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })
  
  it('calls handleChange when e-mail is changed', () => {
    const spy = spyOn(wrapper.instance(), 'handleChange')
    wrapper.instance().forceUpdate()
    const mockEvent = { target: { value: 'something@gmail.com' } }
    wrapper.find('.email').simulate('change', mockEvent)
    expect(spy).toHaveBeenCalled()
  })
  
  it('calls handleChange when password is changed', () => {
    const spy = spyOn(wrapper.instance(), 'handleChange')
    wrapper.instance().forceUpdate()
    const mockEvent = { target: { value: 'password' } }
    wrapper.find('.password').simulate('change', mockEvent)
    expect(spy).toHaveBeenCalled()
    
  })
  
  it('updates state when handleChange is called', () => {
    const mockEvent = { target: { name: 'email', value: 'something@gmail.com' } }
    wrapper.instance().handleChange(mockEvent)
    expect(wrapper.state('email')).toBe('something@gmail.com')
  })
  
  it('calls handleSubmit onSubmit of the form', () => { 
    const spy = spyOn(wrapper.instance(), 'handleSubmit');
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.instance().forceUpdate();
    wrapper.find('form').simulate('submit', mockEvent)
    expect(spy).toHaveBeenCalled()
  })
  
  describe('handleSubmit', () => {
    const mockEvent = {preventDefault: jest.fn()}
    let wrapper;
    let loginUserMock;

    beforeEach(() => {
      loginUserMock = jest.fn()
      wrapper = shallow(<Login loginUser={loginUserMock} />)
      wrapper.state().email = 'ashley@gmail.com'
      wrapper.state().password = 'jake'
    })

    it('Should call login user from props', async () => {
      const expected = {"email": "ashley@gmail.com", "password": "jake"}

      await wrapper.instance().handleSubmit(mockEvent)
      expect(loginUserMock).toHaveBeenCalledWith(expected)      
    })

    it('Should set valid user to true if everything is ok', async () => {
      await wrapper.instance().handleSubmit(mockEvent)

      const expected = true;

      expect(wrapper.state().validUser).toBe(expected)
    })

    it('Should set incorrectlogin to true if there is an error', async () => {
      loginUserMock = () => {throw new Error()}
      wrapper = shallow(<Login loginUser={loginUserMock} />)
      
      await wrapper.instance().handleSubmit(mockEvent)

      const expected = true;

      expect(wrapper.state().incorrectLogin).toBe(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('calls dispatch with a signIn action when loginUser is called', async () => {

      
      const mockDispatch = jest.fn()
      const actionToDispatch = await loginUser({name: 'Justin', email: 'J@aol.com', id: 1, password: 'password'})
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.loginUser({name: 'Justin', email: 'J@aol.com', id: 1, password: 'password'})
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
  
}) 
