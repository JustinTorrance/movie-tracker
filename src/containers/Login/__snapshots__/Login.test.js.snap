// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    loginUser={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="form-title"
        >
          movie
          <img
            alt="movie roll"
            className="form-logo"
            src="./images/movie_roll.svg"
          />
          tracker
        </h1>,
        <h2
          className="login-title"
        >
          Login
        </h2>,
        <form
          className="login-form"
          onSubmit={[Function]}
        >
          <input
            className="email form-input false"
            name="email"
            onChange={[Function]}
            placeholder="email"
            type="text"
            value=""
          />
          <input
            className="password form-input false"
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
          <button>
            Login
          </button>
        </form>,
        <div
          className="new-account-container"
        >
          <h3
            className="incorrect-login false"
          >
            username or password is incorrect
          </h3>
          <h3>
            Don't have an account?
          </h3>
          <Link
            className="form-link"
            replace={false}
            to="/signup"
          >
            Create New Account
          </Link>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "movie",
            <img
              alt="movie roll"
              className="form-logo"
              src="./images/movie_roll.svg"
            />,
            "tracker",
          ],
          "className": "form-title",
        },
        "ref": null,
        "rendered": Array [
          "movie",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "movie roll",
              "className": "form-logo",
              "src": "./images/movie_roll.svg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "tracker",
        ],
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login",
          "className": "login-title",
        },
        "ref": null,
        "rendered": "Login",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="email form-input false"
              name="email"
              onChange={[Function]}
              placeholder="email"
              type="text"
              value=""
            />,
            <input
              className="password form-input false"
              name="password"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />,
            <button>
              Login
            </button>,
          ],
          "className": "login-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "email form-input false",
              "name": "email",
              "onChange": [Function],
              "placeholder": "email",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "password form-input false",
              "name": "password",
              "onChange": [Function],
              "placeholder": "password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Login",
            },
            "ref": null,
            "rendered": "Login",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3
              className="incorrect-login false"
            >
              username or password is incorrect
            </h3>,
            <h3>
              Don't have an account?
            </h3>,
            <Link
              className="form-link"
              replace={false}
              to="/signup"
            >
              Create New Account
            </Link>,
          ],
          "className": "new-account-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "username or password is incorrect",
              "className": "incorrect-login false",
            },
            "ref": null,
            "rendered": "username or password is incorrect",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Don't have an account?",
            },
            "ref": null,
            "rendered": "Don't have an account?",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Create New Account",
              "className": "form-link",
              "replace": false,
              "to": "/signup",
            },
            "ref": null,
            "rendered": "Create New Account",
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="form-title"
          >
            movie
            <img
              alt="movie roll"
              className="form-logo"
              src="./images/movie_roll.svg"
            />
            tracker
          </h1>,
          <h2
            className="login-title"
          >
            Login
          </h2>,
          <form
            className="login-form"
            onSubmit={[Function]}
          >
            <input
              className="email form-input false"
              name="email"
              onChange={[Function]}
              placeholder="email"
              type="text"
              value=""
            />
            <input
              className="password form-input false"
              name="password"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />
            <button>
              Login
            </button>
          </form>,
          <div
            className="new-account-container"
          >
            <h3
              className="incorrect-login false"
            >
              username or password is incorrect
            </h3>
            <h3>
              Don't have an account?
            </h3>
            <Link
              className="form-link"
              replace={false}
              to="/signup"
            >
              Create New Account
            </Link>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "movie",
              <img
                alt="movie roll"
                className="form-logo"
                src="./images/movie_roll.svg"
              />,
              "tracker",
            ],
            "className": "form-title",
          },
          "ref": null,
          "rendered": Array [
            "movie",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "movie roll",
                "className": "form-logo",
                "src": "./images/movie_roll.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "tracker",
          ],
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
            "className": "login-title",
          },
          "ref": null,
          "rendered": "Login",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="email form-input false"
                name="email"
                onChange={[Function]}
                placeholder="email"
                type="text"
                value=""
              />,
              <input
                className="password form-input false"
                name="password"
                onChange={[Function]}
                placeholder="password"
                type="password"
                value=""
              />,
              <button>
                Login
              </button>,
            ],
            "className": "login-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "email form-input false",
                "name": "email",
                "onChange": [Function],
                "placeholder": "email",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "password form-input false",
                "name": "password",
                "onChange": [Function],
                "placeholder": "password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Login",
              },
              "ref": null,
              "rendered": "Login",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3
                className="incorrect-login false"
              >
                username or password is incorrect
              </h3>,
              <h3>
                Don't have an account?
              </h3>,
              <Link
                className="form-link"
                replace={false}
                to="/signup"
              >
                Create New Account
              </Link>,
            ],
            "className": "new-account-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "username or password is incorrect",
                "className": "incorrect-login false",
              },
              "ref": null,
              "rendered": "username or password is incorrect",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Don't have an account?",
              },
              "ref": null,
              "rendered": "Don't have an account?",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Create New Account",
                "className": "form-link",
                "replace": false,
                "to": "/signup",
              },
              "ref": null,
              "rendered": "Create New Account",
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Login Component should match the snapshot when user is logged in 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    loginUser={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "push": false,
      "to": "/",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "push": false,
        "to": "/",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
